bump janet version to 1.31.0

AlecTroemel opened the pull request:
this includes the work by @sogaiu to close issue https://github.com/nesbox/TIC-80/issues/2285.

Full release changes:
https://github.com/janet-lang/janet/releases/tag/v1.31.0
https://github.com/janet-lang/janet/releases/tag/v1.30.0

sogaiu commented:
@AlecTroemel Ok, I tried changing `state.c` a bit and attempted a local Windows build.  That seems to have worked.

The change I made was:

```diff
diff --git a/src/core/state.c b/src/core/state.c
index 4c976da..5f34b07 100644
--- a/src/core/state.c
+++ b/src/core/state.c
@@ -27,6 +27,10 @@
 #include "util.h"
 #endif
 
+#ifdef JANET_WINDOWS
+#include "windows.h"
+#endif
+
 JANET_THREAD_LOCAL JanetVM janet_vm;
 
 JanetVM *janet_local_vm(void) {
```

I don't know if that's really an appropriate change (e.g. may be `winnt.h` would be better?), but at least I get a working `tic80.exe`.

When you get a chance, would you mind updating the PR to see if the CI stuff works?

---

Note that `build/janet/janetconf.h`'s values seem off -- some of [these](https://github.com/nesbox/TIC-80/blob/8c5f8b75e290378a6110f54a0d8d8471942b5a1b/build/janet/janetconf.h#L7-L10).  So may be it makes sense to tweak?

sogaiu commented:
@AlecTroemel `build/janet/janetconf.h`'s values seem off -- specifically some of [these](https://github.com/nesbox/TIC-80/blob/8c5f8b75e290378a6110f54a0d8d8471942b5a1b/build/janet/janetconf.h#L7-L10).  See below for specifics.

Also, Janet's `janetconf.h` in the 1.31.0 release had slightly off values too ([a PR for that was merged](https://github.com/janet-lang/janet/pull/1292)), so I think it might make sense to update to the commit that has those changes.

In summary, I believe I was able to build under Windows locally [1] with the following changes:

* Update `build/janet/janetconf.h`'s values (see the right side of [this](https://github.com/janet-lang/janet/pull/1292/commits/4f642c0843f7b3a5fa44190df00ecc41f156d622) for what I tried)
* Update `vendor/janet` to [this commit of Janet](https://github.com/janet-lang/janet/commit/34767f1e13823114ba5d8e289dbfbb2521db8ba2)

I'll give this another try from scratch and report back.

---

[1] It was nice to have [these notes](https://github.com/nesbox/TIC-80/pull/2244#issuecomment-1661699227).  The footnote in that comment was handy too ;)

AlecTroemel commented:
@sogaiu i've updated the `janetconf.h`. I've also updated janet to [this commit](https://github.com/janet-lang/janet/commit/169bd812c9183865b2399a8eaeab6aad238856c5), since it included the windows fix you mentioned in [this issue](https://github.com/janet-lang/janet/issues/1293)
